### DISCUSSION TITLE
Dynamic plot vs. static SVG from performance perspective

### DISCUSSION THREAD

**Discussion by qui8t** [facebook/docusaurus#9434] [Q&A] [answered]
I have ~10 plots in my Docusaurus-generated website. I have two options for displaying them in `.mdx` files:

1. Plot the 10 data sources using any library and generate SVG/PNG figures, and use the static figures in the website;
2. Put the 10 data sources as static resources on the website, write a custom script to parse the data sources and plot them using Plotly.JS, and call the script from the `.mdx` files.

Each of these options have ups and downs, e.g., option 1 gives the freedom of plotting data in any library, possibly with more options than plotly, and option 2 provides more interactive experience with the plot (e.g., zoom, pan, select). 

I want to learn about their difference from the website performance perspective; specifically:

- Assuming 10 CSV files containing the data, each of size ~1MB and data plotted on different `.mdx` files; does it mean every time the website is opened, 10MB worth of data is copied to the client and all the plots are re-generated regardless of which page the user views? 

- Should I expect significant slow-down of the website when using the second option?

**Comment by Josh-Cena** [comment, marked as answer]
> Assuming 10 CSV files containing the data, each of size ~1MB and data plotted on different .mdx files; does it mean every time the website is opened, 10MB worth of data is copied to the client and all the plots are re-generated regardless of which page the user views?

That depends on your implementation, but I would expect any straightforward one to only download the data when the page in question is actually opened.

> Should I expect significant slow-down of the website when using the second option?

Interactivity always comes at a cost, so in this case I would say it's fine, especially if you provide a proper load spinner. In fact, why don't you do both: display a prerendered SVG while the heavier data is loading? You can take a look at our `BrowserOnly` component.

**Comment by qui8t** [comment]
> I would expect any straightforward one to only download the data when the page in question is actually opened.

> especially if you provide a proper load spinner.

Would you please elaborate? A simplified version of my implementation is as the following, any suggestions on how best I can update it?

```javascript
import Plot from 'react-plotly.js';
import Papa from 'papaparse';

function Plot() {
  const path = useBaseUrl('/data.csv')

  useEffect(() => {
    fetch(path)
      // parse the csv
     setData(parsedData);
  }, []);

  if (data)
    return (
      <Plot
        layout={{ title: 'Plot' }}
        data={[
          {
            y: data[0],
            type: 'box',
          }
        ]}
      />
    );
}

export default Plot;
```

which I call it in the markdowns as the following. 

```
---
title: Some plots
---

import Plot from "../../src/components/plot.js"

<Plot />
```

**Reply by Josh-Cena** [reply]
Yep, that seems like a good implementation to me. In this case, the component will only be executed on this particular page, which means the fetch won't happen on unrelated pages. (And even if it does, it's not render-blocking because you put it in an effect, so it's still good user experience). After `if (data) return <...>`, you can return another fallback image, such as `<svg>Your prerendered SVG</svg>`, or a simple CSS-based load spinner, like this: https://github.com/facebook/docusaurus/blob/main/packages/docusaurus/src/client/theme-fallback/Loading/index.tsx

**Reply by qui8t** [reply]
That is perfect, thank you!!
